plugins/microservices/msi_tests/libmsi_test_scoped_permission.cpp

```diff
@@ -0,0 +1,274 @@
+/// \file
+
+#include "dstream.hpp"
+#include "getRodsEnv.h"
+#include "irods_exception.hpp"
+#include "irods_ms_plugin.hpp"
+#include "irods_re_structs.hpp"
+#include "msParam.h"
+#include "msi_assertion_macros.hpp"
+#include "scoped_permission.hpp"
+
+#define IRODS_IO_TRANSPORT_ENABLE_SERVER_SIDE_API
+#include "transport/default_transport.hpp"
+
+#define IRODS_FILESYSTEM_ENABLE_SERVER_SIDE_API
+#include "filesystem.hpp"
+
+namespace
+{
+    namespace fs = irods::experimental::filesystem;
+    namespace io = irods::experimental::io;
+    using sp = irods::experimental::scoped_permission;
+
+    auto collection_path(RsComm* comm) -> fs::path
+    {
+        fs::path p = "/";
+        p /= comm->myEnv.rodsZone; // NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+        p /= "home";
+        p /= comm->clientUser.userName; // NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+
+        return p;
+    } // collection_path
+
+    auto make_test_object(RsComm* comm) -> fs::path
+    {
+        const auto path = collection_path(comm) / "test_scoped_permission.txt";
+
+        const std::string_view contents = "testdata";
+
+        // Create a new data object using the server-side API.
+        {
+            io::server::native_transport tp{*comm};
+            io::odstream{tp, path} << contents;
+        }
+
+        return path;
+    } // make_test_object
+
+    auto cleanup(RsComm* comm, const fs::path& path) -> void
+    {
+        if (fs::server::exists(*comm, path)) {
+            // Ironically, we must use scoped permissions to ensure the ability to remove the object.
+            const auto temporary_ownership = sp{*comm, path, fs::perms::own};
+            fs::server::remove(*comm, path, fs::remove_options::no_trash);
+        }
```

Comment:
So this is the case where the perms cannot be restored due to the target object being removed.

Reply:
Correct. Everything works as expected, you just get some errors on the way out the door. Something to consider.